### ISSUE TITLE
Sitemap SSE connection closes prematurely when item state changes

### ISSUE THREAD

**Issue by mhilbush** [eclipse/smarthome#4871]
I'm seeing an issue with my sitemap on one of my systems.  Every time the CurrentTime item updates (which is every minute), in BasicUI I get the message **Offline: waiting for connection to become available**

I have an item that links to the NTP dateTime channel
```
DateTime CurrentTime "Date [%1$tA, %1$tm/%1$td/%1$tY %1$tT]" <calendar> { channel="ntp:ntp:local:dateTime" }
```

And I have a sitemap that looks like this
```
sitemap test label="Test"
{
    Frame {
        Text item=CurrentTime  label="Today is [%1$tA, %1$tm/%1$td/%1$tY %1$tl:%1$tM %1$tp]"       icon="calendar"
    }
}
```

Both systems have similar configs;
- OH build 1140
- Ubuntu Linux kernel 4.4.0-101-generic
- openjdk version "1.8.0_151"

I've enabled DEBUG logging of `org.eclipse.smarthome.io.rest.sitemap`

For the system that's working correctly, I see these log entries in my openhab.log
```
2018-01-07 09:00:18.319 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 6545592a-69f4-4c09-a7a8-60e7b77772fc
2018-01-07 09:00:18.413 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 6545592a-69f4-4c09-a7a8-60e7b77772fc changed to page 43p of sitemap 43p
2018-01-07 09:00:18.415 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 6545592a-69f4-4c09-a7a8-60e7b77772fc.
2018-01-07 09:01:04.316 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 6545592a-69f4-4c09-a7a8-60e7b77772fc.
2018-01-07 09:02:04.314 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 6545592a-69f4-4c09-a7a8-60e7b77772fc.
2018-01-07 09:03:04.314 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 6545592a-69f4-4c09-a7a8-60e7b77772fc.
2018-01-07 09:04:04.316 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 6545592a-69f4-4c09-a7a8-60e7b77772fc.
2018-01-07 09:05:04.319 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 6545592a-69f4-4c09-a7a8-60e7b77772fc.
2018-01-07 09:06:04.320 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 6545592a-69f4-4c09-a7a8-60e7b77772fc.
```

For the system that's not working correctly, I see this in my openhab.log.  The **SSE connection closed** log entry occurs at the same time as the `CurrentTime` item update is logged in event.log.  When this occurs, Basic UI presents the **Offline: waiting for connection to become available**.
```
2018-01-07 09:01:25.948 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 480fe861-bbda-4432-86f5-a2880bc4af38
2018-01-07 09:01:26.085 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 480fe861-bbda-4432-86f5-a2880bc4af38 changed to page test of sitemap test
2018-01-07 09:01:26.085 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 480fe861-bbda-4432-86f5-a2880bc4af38.
2018-01-07 09:02:15.907 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription 480fe861-bbda-4432-86f5-a2880bc4af38 has been closed.
2018-01-07 09:02:15.908 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id 480fe861-bbda-4432-86f5-a2880bc4af38
2018-01-07 09:02:25.950 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id b19ec227-71c7-47c5-af80-ef35621717e9
2018-01-07 09:02:26.088 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription b19ec227-71c7-47c5-af80-ef35621717e9 changed to page test of sitemap test
2018-01-07 09:02:26.089 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription b19ec227-71c7-47c5-af80-ef35621717e9.
2018-01-07 09:03:15.908 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription b19ec227-71c7-47c5-af80-ef35621717e9 has been closed.
2018-01-07 09:03:15.908 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id b19ec227-71c7-47c5-af80-ef35621717e9
2018-01-07 09:03:25.948 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 64d2d11f-2ea2-4c8a-a012-71f2dd729f0f
2018-01-07 09:03:26.078 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 64d2d11f-2ea2-4c8a-a012-71f2dd729f0f changed to page test of sitemap test
2018-01-07 09:03:26.078 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 64d2d11f-2ea2-4c8a-a012-71f2dd729f0f.
2018-01-07 09:04:15.908 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription 64d2d11f-2ea2-4c8a-a012-71f2dd729f0f has been closed.
2018-01-07 09:04:15.909 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id 64d2d11f-2ea2-4c8a-a012-71f2dd729f0f
2018-01-07 09:04:25.952 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id d2719be6-b047-4d1d-90c5-5dbef28d43b5
2018-01-07 09:04:26.090 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription d2719be6-b047-4d1d-90c5-5dbef28d43b5 changed to page test of sitemap test
2018-01-07 09:04:26.091 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription d2719be6-b047-4d1d-90c5-5dbef28d43b5.
2018-01-07 09:05:15.909 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription d2719be6-b047-4d1d-90c5-5dbef28d43b5 has been closed.
2018-01-07 09:05:15.909 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id d2719be6-b047-4d1d-90c5-5dbef28d43b5
```
In the browser console, I see network behavior that looks like this.
![8afbec7067e2e99c5c339f980dd7d87763c48915_1_690x308](https://user-images.githubusercontent.com/19286354/34654025-a8a71acc-f3c3-11e7-8286-ed6e75fdd284.jpg)

We are speculating that the Cache-Control header should be set to no-cache in the server response, based on some reports on the Internet; however, we are uncertain of this and would like to get additional input.

@lolodomo has been helping me troubleshoot.  Additional info on our troubleshooting steps can be found in this community thread.
https://community.openhab.org/t/no-autoupdate-in-basicui-after-sitemap-update/37726/9

**Comment by lolodomo** [comment]
Unfortunately this issue is not reproduceable by everyone, at least by me.

By the way, all my readings of yesterday about SSE show that the server should set Control-Cache header to "no-cache". I think we could try to apply this change and see with @mhilbush if it helps.
Other point not clear is whether chunked transfer encoding is a problem or not.
https://www.w3.org/TR/eventsource/#notes

**Comment by mhilbush** [comment]
I forgot to mention that this problem is reproducible on a variety of client platforms.
- Windows (Chrome and Firefox)
- Mac (Chrome, Firefox, Safari)
- Linux (Chrome and Firefox)

**Comment by mhilbush** [comment]
On the system that's not working correctly, I changed the NTP refresh interval from 60 to 10.  I now see activity in openhab.log that looks like this.  So the sitemap events are definitely being sent. However, the connection still is closing periodically. I guess it's time for wireshark.  :-(
```
2018-01-09 11:45:57.561 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription ccab3686-1e32-4a06-88c3-59dd4ddbf79c.
2018-01-09 11:46:07.339 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription ccab3686-1e32-4a06-88c3-59dd4ddbf79c.
2018-01-09 11:46:17.360 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription ccab3686-1e32-4a06-88c3-59dd4ddbf79c.
2018-01-09 11:46:27.361 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription ccab3686-1e32-4a06-88c3-59dd4ddbf79c.
2018-01-09 11:46:37.361 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription ccab3686-1e32-4a06-88c3-59dd4ddbf79c.
2018-01-09 11:46:47.363 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription ccab3686-1e32-4a06-88c3-59dd4ddbf79c has been closed.
2018-01-09 11:46:47.363 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id ccab3686-1e32-4a06-88c3-59dd4ddbf79c
2018-01-09 11:46:57.476 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 0b99aa83-7644-4942-b903-27ea52d5e788
2018-01-09 11:46:57.566 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 0b99aa83-7644-4942-b903-27ea52d5e788 changed to page test of sitemap test
2018-01-09 11:46:57.566 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 0b99aa83-7644-4942-b903-27ea52d5e788.
2018-01-09 11:47:07.364 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 0b99aa83-7644-4942-b903-27ea52d5e788.
2018-01-09 11:47:17.364 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 0b99aa83-7644-4942-b903-27ea52d5e788.
2018-01-09 11:47:27.364 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 0b99aa83-7644-4942-b903-27ea52d5e788.
2018-01-09 11:47:37.365 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 0b99aa83-7644-4942-b903-27ea52d5e788.
2018-01-09 11:47:47.366 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription 0b99aa83-7644-4942-b903-27ea52d5e788 has been closed.
2018-01-09 11:47:47.366 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id 0b99aa83-7644-4942-b903-27ea52d5e788
2018-01-09 11:47:57.404 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 0bdbb524-030e-480b-a595-1639c398050b
2018-01-09 11:47:57.499 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 0bdbb524-030e-480b-a595-1639c398050b changed to page test of sitemap test
2018-01-09 11:47:57.500 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 0bdbb524-030e-480b-a595-1639c398050b.
2018-01-09 11:48:07.367 [DEBUG] [.sitemap.internal.SitemapEventOutput] - Sent sitemap event for widget 0000 to subscription 0bdbb524-030e-480b-a595-1639c398050b.
```

**Comment by lolodomo** [comment]
Interesting it is not at each event.

**Comment by mhilbush** [comment]
Yes, it is.

I turned on more debugging and I'm starting to think it has nothing to to with the events.

I see this repeated every minute in my logs, and I have no idea what to make of it.
```
2018-01-09 14:20:13.127 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 10531f04-bfce-443b-b3c3-27c8e10f3ad8
2018-01-09 14:20:13.183 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 10531f04-bfce-443b-b3c3-27c8e10f3ad8 changed to page test of sitemap test
2018-01-09 14:20:13.184 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 10531f04-bfce-443b-b3c3-27c8e10f3ad8.
2018-01-09 14:20:30.010 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STOPPING - swagger-jersey2-jaxrs
2018-01-09 14:20:30.011 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STOPPED - swagger-jersey2-jaxrs
2018-01-09 14:20:30.015 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent UNRESOLVED - swagger-jersey2-jaxrs
2018-01-09 14:20:30.029 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent RESOLVED - swagger-jersey2-jaxrs
2018-01-09 14:20:30.037 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STARTING - swagger-jersey2-jaxrs
2018-01-09 14:20:30.038 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STARTED - swagger-jersey2-jaxrs
2018-01-09 14:20:45.890 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription 10531f04-bfce-443b-b3c3-27c8e10f3ad8 has been closed.
2018-01-09 14:20:45.890 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id 10531f04-bfce-443b-b3c3-27c8e10f3ad8
2018-01-09 14:20:55.931 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 27946f85-bf67-40a9-8205-f897afa02fb0
2018-01-09 14:20:56.071 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 27946f85-bf67-40a9-8205-f897afa02fb0 changed to page test of sitemap test
2018-01-09 14:20:56.072 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 27946f85-bf67-40a9-8205-f897afa02fb0.
2018-01-09 14:21:30.217 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STOPPING - swagger-jersey2-jaxrs
2018-01-09 14:21:30.218 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STOPPED - swagger-jersey2-jaxrs
2018-01-09 14:21:30.224 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent UNRESOLVED - swagger-jersey2-jaxrs
2018-01-09 14:21:30.241 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent RESOLVED - swagger-jersey2-jaxrs
2018-01-09 14:21:30.250 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STARTING - swagger-jersey2-jaxrs
2018-01-09 14:21:30.250 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STARTED - swagger-jersey2-jaxrs
2018-01-09 14:21:45.887 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription 27946f85-bf67-40a9-8205-f897afa02fb0 has been closed.
2018-01-09 14:21:45.887 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id 27946f85-bf67-40a9-8205-f897afa02fb0
2018-01-09 14:21:55.929 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 87eab94c-f37e-4824-ae7d-1092e164b7ab
2018-01-09 14:21:56.051 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 87eab94c-f37e-4824-ae7d-1092e164b7ab changed to page test of sitemap test
2018-01-09 14:21:56.052 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 87eab94c-f37e-4824-ae7d-1092e164b7ab.
2018-01-09 14:22:30.076 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STOPPING - swagger-jersey2-jaxrs
2018-01-09 14:22:30.077 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STOPPED - swagger-jersey2-jaxrs
2018-01-09 14:22:30.082 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent UNRESOLVED - swagger-jersey2-jaxrs
2018-01-09 14:22:30.093 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent RESOLVED - swagger-jersey2-jaxrs
2018-01-09 14:22:30.101 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STARTING - swagger-jersey2-jaxrs
2018-01-09 14:22:30.101 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STARTED - swagger-jersey2-jaxrs
2018-01-09 14:22:45.887 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription 87eab94c-f37e-4824-ae7d-1092e164b7ab has been closed.
2018-01-09 14:22:45.888 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id 87eab94c-f37e-4824-ae7d-1092e164b7ab
2018-01-09 14:22:56.216 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Created new subscription with id 24397bb2-1f79-4433-943f-30e34e55950d
2018-01-09 14:22:56.336 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Subscription 24397bb2-1f79-4433-943f-30e34e55950d changed to page test of sitemap test
2018-01-09 14:22:56.337 [DEBUG] [est.sitemap.internal.SitemapResource] - Client requested sitemap event stream for subscription 24397bb2-1f79-4433-943f-30e34e55950d.
2018-01-09 14:23:29.952 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STOPPING - swagger-jersey2-jaxrs
2018-01-09 14:23:29.953 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STOPPED - swagger-jersey2-jaxrs
2018-01-09 14:23:29.958 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent UNRESOLVED - swagger-jersey2-jaxrs
2018-01-09 14:23:29.972 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent RESOLVED - swagger-jersey2-jaxrs
2018-01-09 14:23:29.980 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STARTING - swagger-jersey2-jaxrs
2018-01-09 14:23:29.981 [DEBUG] [swagger-jersey2-jaxrs               ] - BundleEvent STARTED - swagger-jersey2-jaxrs
2018-01-09 14:23:45.888 [DEBUG] [est.sitemap.internal.SitemapResource] - SSE connection for subscription 24397bb2-1f79-4433-943f-30e34e55950d has been closed.
2018-01-09 14:23:45.888 [DEBUG] [t.sitemap.SitemapSubscriptionService] - Removed subscription with id 24397bb2-1f79-4433-943f-30e34e55950d
```

**Comment by maggu2810** [comment]
Hm, could this be caused by some openHAB Karaf integration bundle that e.g. triggers some bundle / feature handling?
Can you try to reproduce the error without any openHAB logic running in the background and using e.g. the ESH launch configuration of the official ESH Eclipse IDE setup?

**Comment by mhilbush** [comment]
Ah, so the SSE issue is just the victim.  Good idea!

I don't have the IDE on this box.  And, it's my only OH instance that's exhibiting this problem.

I only install bindings through services/addons.cfg.  So, one by one I deleted the installed bindings.  Started at the top (I left package=minimal) and worked my way down, I finally got to the misc line, which had
```
misc = gcal1,restdocs
```
Deleted gcal1 -- nothing

Deleted restdocs -- the problem went away!

Added restdocs -- problem did NOT return.

Added all my other bindings, and it appears to be functioning normally again.

I have no explanation, but I'm going to close this issue.
@lolodomo Thanks for your time the other day
@maggu2810 Thanks for your suggestion which pointed me in the right direction.

**Comment by mhilbush** [comment]
Just to confirm it was the removal of restdocs that fixed it:
- I restored the backup I had made just before starting the process described above
- I removed ONLY restdocs from addons.cfg
- Confirmed problem went away
- Readded restdocs to addons.cfg
- Confirmed problem did not reoccur

Is there anything I should look at in my backup to help better explain why this was happening?

**Comment by maggu2810** [comment]
You could try to identify "why" the bundle has been stopped and started all the time.
But for me it seems to be related to the openHAB Karaf integration and so not really related to the Eclipse SmartHome framework itself.
If Kai is interested in (to prevent that in other instances of the oh-distro) perhaps an issue in a openHAB tracker should be created. You could then set all your log levels (or perhaps only the openHAB ones, then only the Karaf ones, then only the OSGi core and DS / SCR ones) to info, then to debug and then to trace. Perhaps you can identify the "trigger".

**Comment by kaikreuzer** [comment]
As the bundle re-installation happens once a minute, I very much suspect that it is related to the Karaf Feature installer in openhab-core.
@mhilbush If you can easily reproduce this behavior, please file an issue at https://github.com/openhab/openhab-core/issues for it - thanks!

**Comment by mhilbush** [comment]
I'll open an issue.

I didn't mention this, but earlier I set DEBUG log level on `org.openhab.core.karaf`. I saw the reinstallation occur when the feature installer ran.  That would seem to confirm your suspicion.

**Comment by mhilbush** [comment]
@kaikreuzer I never did open an issue for this.  From what I can tell, the sitemap repeatedly going offline was just the victim of another problem.

I tried for quite a few hours to track down what was causing the problem.  I think I can definitively say that the issue had nothing to do with the item state changing, as the title of this issue suggests.  Therefore I'll edit the title to something more accurate in order to avoid any confusion.

Furthermore, of my 4 environments (prod1, prod2, test, and dev), this problem occurred only on my test environment.  I could not reproduce it on any other environment.  Yet, while the problem could be reproduced consistently in the test environment, I was unable to determine what was causing the behavior. 

I looked at the system configuration before I removed restdocs from addons.cfg, as well as after it was removed from addons.cfg.  Other than some changes to the internal configuration files deep in the cache directory (which I don't understand at all), I saw nothing else noteworthy in what changed between the two tests.  OTOH, I do find it odd that the problem persisted after a restart even when removing the contents of tmp and cache.

The problem has not reoccurred since.  I can only assume that there was something corrupted in my configuration.

**Comment by mhilbush** [comment]
@kaikreuzer I completely agree that this should be set aside. Weird -- I can't think of a better word to describe it!

In the course of my analysis, and unrelated to this issue, I did see some unusual memory behavior in my sar logs.  Rather than continue to add to this thread, I'm going to open a separate issue, probably in openhab/core.
